src/cordic.rs

```diff
@@ -0,0 +1,915 @@
+//!     let mut rcc = dp.RCC.freeze(Config::hsi(), pwr);
+//!
+//!     let mut cordic = dp
+//!         .CORDIC
+//!         .constrain(&mut rcc)
+//!         .freeze::<I1F15, I1F31, SinCos, P60>(); // 16 bit arguments, 32 bit results, compute sine and cosine, 60 iterations
+//!
+//!     // static operation (zero overhead)
+//!
+//!     cordic.start(I1F15::from_num(-0.25 /* -45 degreees */));
+//!
+//!     let (sin, cos) = cordic.result();
+//!
+//!     println!("sin: {}, cos: {}", sin.to_num::<f32>(), cos.to_num::<f32>());
+//!
+//!     // dynamic operation
+//!
+//!     let mut cordic = cordic.into_dynamic();
+//!
+//!     let sqrt = cordic.run::<Sqrt<N0>>(I1F15::from_num(0.25));
+//!     println!("sqrt: {}", sqrt.to_num::<f32>());
+//!     let magnitude = cordic.run::<Magnitude>((I1F15::from_num(0.25), I1F15::from_num(0.5)));
+//!     println!("magnitude: {}", magnitude.to_num::<f32>());
+//!
+//!     loop {}
+//! }
+//! ```
+
+use crate::{rcc::Rcc, stm32::CORDIC};
+use core::marker::PhantomData;
+
+/// Extension trait for constraining the CORDIC peripheral.
+pub trait Ext {
+    /// Constrain the CORDIC peripheral.
+    fn constrain(self, rcc: &mut Rcc) -> CordicReset;
+}
+
+impl Ext for CORDIC {
+    #[inline]
+    fn constrain(self, rcc: &mut Rcc) -> CordicReset {
+        rcc.rb.ahb1enr.modify(|_, w| w.cordicen().set_bit());
+
+        // SAFETY: the resource is assumed to be
+        // in the "reset" configuration.
+        unsafe { Cordic::wrap(self) }
+    }
+}
+
+/// Traits and structures related to data types.
+pub mod types {
+    use fixed::traits::Fixed;
+
+    pub use fixed::types::{I1F15, I1F31};
+
+    /// Trait for newtypes to represent CORDIC argument or result data.
+    pub trait DataType: Fixed {
+        /// Convert to bits of the register width,
+        fn to_register(self) -> u32;
+        /// Convert from bits of the register width,
+        fn from_register(bits: u32) -> Self;
+    }
+
+    impl DataType for I1F31 {
+        fn to_register(self) -> u32 {
+            self.to_bits() as u32
+        }
+        fn from_register(bits: u32) -> Self {
+            Self::from_bits(bits as i32)
+        }
+    }
+
+    impl DataType for I1F15 {
+        fn to_register(self) -> u32 {
+            self.to_bits() as u16 as u32
+        }
+        fn from_register(bits: u32) -> Self {
+            Self::from_bits(bits as u16 as i16)
+        }
+    }
+
+    /// Traits and structures related to argument type-states.
+    pub mod arg {
+        pub(crate) type Raw = crate::stm32::cordic::csr::ARGSIZE_A;
+
+        /// Trait for argument type-states.
+        pub(crate) trait State: super::DataType {
+            /// Raw representation of the configuration
+            /// in the form of a bitfield variant.
+            const RAW: Raw;
+
+            /// Configure the resource to be represented
+            /// by this type-state.
+            #[inline]
+            fn set<const O: u8>(w: crate::stm32::cordic::csr::ARGSIZE_W<O>) {
```

Comment:
(Feel free to ignore)

Maybe it's just me but, `O` looks a hole lot like a zero `0` to me

Reply:
Yeah I just copied the PAC, I think it stands for `Offset`, I'll change it to that.